Sources/SnapshotTesting/Snapshotting/UIImage.swift

```diff
@@ -76,7 +78,39 @@ let imageContextColorSpace = CGColorSpace(name: CGColorSpace.sRGB)
 let imageContextBitsPerComponent = 8
 let imageContextBytesPerPixel = 4
 
-private func compare(_ old: UIImage, _ new: UIImage, precision: Float) -> Bool {
+/// A comparison strategy for snapshots.
+public enum ComparisonStrategy {
+  /**
+   The difference is calculated as the percentage of bytes that are not equal.
+   
+   let difference = (0..<byteCount).reduce(0) { r, i in r += old[i] != new[i] ? 1 : 0 }
+   */
+  case equality
+  /**
+   The difference is calculated as the sum of normalized differences between the old and new values taken modulo.
+   
+   let difference = (0..<byteCount).reduce(0) { r, i in r += abs(old[i] - new[i]) / 255 }
+   */
+  case subtractionModule
+  public typealias ComparisonClosure = (_ oldByte: UInt8, _ newByte: UInt8) -> Float
+  /**
+   The difference is calculated as the sum of the values returned from the comparison closure performed for each pair of bytes.
+   */
+  case custom(ComparisonClosure)
+  
+  var closure: ComparisonClosure {
+    switch self {
+    case .equality:
+      return { ($0 == $1) ? 0 : 1 }
+    case .subtractionModule:
+      return { abs(Float($0) - Float($1)) / 255 }
+    case .custom(let comparisonClosure):
+      return comparisonClosure
+    }
+  }
+}
+
+private func compare(_ old: UIImage, _ new: UIImage, precision: Float, comparisonStrategy: ComparisonStrategy) -> Bool {
```

Comment:
Is there perhaps an argument to be made where the entire comparison function is more easily swappable? Not just the method in which bytes are compared.

This could allow people to provide their own comparison solutions, or threshold matching, etc., without needing to create their entire own strategy. (it's nice to still reuse all the view loading/sizing/snapshotting mechanisms that this library provides)

Reply:
Hi @Sherlouk, thank you very much for your interest in our pull request!

We originally went the same route as you and added the ability to provide a custom comparison solution via a closure: https://github.com/magic146/swift-snapshot-testing/tree/allow-custom-compare-closure

But then we realized that the comparison of images in almost 100% of cases will be based on a comparison of bitmaps. And having two parameters (the precision and the method in which the bytes are compared) should be enough. In our opinion, the rest of the logic is just a framework for creating an environment for byte comparison, and it will remain the same (or even could be optimised), regardless of the comparison strategy and precision.

Having said that, I have to agree with you that at the moment creating the entire strategy can be a little expensive: while it is feasible to use a custom `Snapshotting`, it is impossible to reuse existing algorithms from the framework. However, we are proposing to make this change a separate pull request as it does not meet our current goals.